### PR TITLE
mgmt: mcumgr: img_mgmt: Prevent re-upload of duplicate image data

### LINKED ISSUE #40924: mgmt/mcumgr/lib: Do not re-upload image, by default, to the secondary slot
**Is your enhancement proposal related to a problem? Please describe.**
Currently every upload request will end up with secondary slot erase and image upload, even if the same image is
already there.
This happens even though the first packet of SMP image upload request caries "sha", that is considered SHA256 sum of an image that is uploaded; even it that is true for client software, that sum is only used to identify current upload session in case it gets broken.
Part of this "hash" is stored in mcumgr volatile state, that does not survive reset, with last successfully uploaded offset, so in case when session gets restarted the mcumgr will match the "hash" with "off" zero packet and if "hash" matches, will request the "off" the state instead of starting from 0.

**Describe the solution you'd like**
The "sha" in packet with "off" 0 should be compared against calculated "sha" of the image in secondary slot and in case when it matches, an application should respond with request for last offset, this way concluding the upload process.
Optional "force" option into upload may be added, to enforce upload even if hash matches, but it generally should be not needed as it is enough to for example binary negate the "sha" to make it fail and yet be usable to identify the session.

The should be configurable by user as it may not be desired for various reasons.
The "sha" algorithm should be user selectable at-compile time as this does not really has to be already assumed SHA256.
Optionally the check may be also done against the "primary" image slot, as uploading, to the secondary slot, image that is already running seems pointless, most of the time.

**Describe alternatives you've considered**
Not doing it, but with Kconfig option added this enhancement will only affect these who desire it.

**Additional context**
Here are lines of mcumgr that handle the "sha" from packet with "off" 0: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c#L443-L479
Clearly there are not calculations done here on existing image.
Example of client side application actually doing the sha calculations: https://github.com/NordicSemiconductor/Android-nRF-Connect-Device-Manager/blob/3a1f9a5c35920530854f66e0009b2c5be5dd5b41/mcumgr-core/src/main/java/io/runtime/mcumgr/managers/ImageManager.java#L226

Note that this feature is currently not supported by mcumgr-cli.

### PR DESCRIPTION
**Pull request by nordicjm**
This adds a hash check when the CONFIG_IMG_ENABLE_IMAGE_CHECK Kconfig option is enabled that will check the underlying image hash to see if it is the same as the one provided by the mcumgr client, and if so, will prevent erasure/uploading the same image data.

Fixes #40924

### CODE REVIEW

**Comment by nordicjm** [comment]
Initial upload with mcumgr:
```
mcumgr --conntype serial --connstring dev=/dev/ttyUSB0,mtu=512 image upload zephyr/zephyr.signed.bin 
 196.15 KiB / 196.15 KiB [===================================================================================] 100.00% 3.08 KiB/s 1m3s
Done
```
Re-upload of same data with mcumgr:
```
mcumgr --conntype serial --connstring dev=/dev/ttyUSB0,mtu=512 image upload zephyr/zephyr.signed.bin 
 196.15 KiB / 196.15 KiB [===================================================================================] 100.00% 279.22 KiB/s 0s
Done
```
Rebooting module and re-uploading same data with mcumgr:
```
mcumgr --conntype serial --connstring dev=/dev/ttyUSB0,mtu=512 reset
Done
mcumgr --conntype serial --connstring dev=/dev/ttyUSB0,mtu=512 image upload zephyr/zephyr.signed.bin 
 196.15 KiB / 196.15 KiB [===================================================================================] 100.00% 279.27 KiB/s 0s
Done
```

File: subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c [hunk 1]
```diff
@@ -432,6 +439,30 @@ img_mgmt_upload(struct mgmt_ctxt *ctxt)
...
 
+#ifdef CONFIG_IMG_ENABLE_IMAGE_CHECK
+		/* Check if the existing image hash matches the hash of the underlying data */
+		if (flash_img_init_id(&ctx, g_img_mgmt_state.area_id) == 0) {
+			struct flash_img_check fic = {
+				.match = g_img_mgmt_state.data_sha,
+				.clen = g_img_mgmt_state.size,
+			};
+
+			flash_area_close(g_img_mgmt_state.area_id);
```
Comment (nvlsianpu) [inline subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:450]: You make visible to me that (maybe) something is weird with usage of flash_img_check(). This call make the confusion to me.

Comment (nordicjm) [reply subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:450]: It's a nuance of using `flash_img_init_id` which leaves a flash area open (which, normally, would be closed by `flash_img_buffered_write` but is not called here so needs to be closed manually. `flash_img_check` does a new flash area open and close

File: subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c [hunk 2]
```diff
@@ -363,6 +367,10 @@ img_mgmt_upload(struct mgmt_ctxt *ctxt)
 	int rc;
 	struct img_mgmt_upload_action action;
 	bool last = false;
+#ifdef CONFIG_IMG_ENABLE_IMAGE_CHECK
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c:370]: Bring this down to `if (req.off == 0) {` block, they are only used there.

Comment (nordicjm) [reply subsys/mgmt/mcumgr/lib/cmd/img_mgmt/src/img_mgmt.c]: Fixed

**Review by de-nordic — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**